Feature: Addition of Venue for Events

wingman47 opened the pull request:
**What kind of change does this PR introduce?**

Introduces venues for events. 
For now, it adds the capability to CRUD venues in organization and their respective tests. 

Subsequent PRs (as requested) will add other features like booking venues for events, avoid clashes of venues, managing event checkins for venue capacity.


**Issue Number:**

Fixes #1640

**Did you add tests for your changes?**

Yes.

**Snapshots/Videos:**

PART - 1

https://github.com/PalisadoesFoundation/talawa-api/assets/49762303/3fa2a010-292c-4894-a440-d30d13c74a7b

PART - 2

https://github.com/PalisadoesFoundation/talawa-api/assets/49762303/756fbbda-dd48-4b87-acfe-fe73b8fb3bdc

**Have you read the [contributing guide](https://github.com/PalisadoesFoundation/talawa-api/blob/master/CONTRIBUTING.md)?**

Yes

Cioppolo14 commented:
@wingman47 Please fix the failing tests, the codecov and introspection tests should pass.

wingman47 commented:
> @wingman47 Please fix the failing tests, the codecov and introspection tests should pass. 

Working on it

wingman47 commented:
> @wingman47 Please fix the failing tests, the codecov and introspection tests should pass.

The failing codecov tests are not related to my changes. 
Only image upload is not tested in my files because I am not sure how to implement it. Any guidance would be helpful.
![image](https://github.com/PalisadoesFoundation/talawa-api/assets/49762303/d881e67c-7e45-4819-9888-dd8bd05077c4)


The workflow errors would be there due to the type of PR I think.

palisadoes commented:
@wingman47 

1. We had to revert a PR recently. Please merge your code with the latest upstream.
2. Try to get the code coverage for you patch to pass

EshaanAgg commented on src/typeDefs/inputs.ts:
```diff
@@ -64,6 +64,15 @@ export const inputs = gql`
 
+  input EditVenueInput {
+    _id: ID!
+    capacity: Int!
+    name: String!
```
These should be allowed to be `undefined`, as it is not compulsory for the user to provide them if they aren't changed.

EshaanAgg commented on src/typeDefs/mutations.ts:
```diff
@@ -111,10 +111,16 @@ export const mutations = gql`
 
     createSampleOrganization: Boolean! @auth
 
+    createVenue(data: VenueInput!, file: String): Venue! @auth
```
What is `file` here? Why isn't it included in the `data` itself?

EshaanAgg commented on src/typeDefs/mutations.ts:
```diff
@@ -111,10 +111,16 @@ export const mutations = gql`
 
     deleteDonationById(id: ID!): DeletePayload!
 
+    deleteVenue(venueId: ID!, organizationId: ID!): Venue @auth
```
No need for `organizationId` here.

wingman47 replied:
`organizationId` is being used here to check here if the user who want to delete the venue is admin of the organization or not. Should I take any other approach?

EshaanAgg commented on src/typeDefs/mutations.ts:
```diff
@@ -111,10 +111,16 @@ export const mutations = gql`
 
+    deleteVenue(venueId: ID!, organizationId: ID!): Venue @auth
+
+    editVenue(data: EditVenueInput!, file: String): Venue! @auth
```
Similar comments about `file`.

EshaanAgg commented on src/typeDefs/types.ts:
```diff
@@ -168,6 +168,7 @@ export const types = gql`
     status: Status!
     feedback: [Feedback!]!
     averageFeedbackScore: Float
+    venue: ID
```
It should be returning a `Venue` with a field-level resolver.

EshaanAgg commented on src/typeDefs/queries.ts:
```diff
@@ -119,5 +119,7 @@ export const queries = gql`
       orderBy: UserOrderByInput
     ): [User]! @auth
+
+    venuesInOrganization(id: ID!): [Venue] @auth
```
No need for this. Add `venues` as a field on the organization model, and then use the field level resolver for the same.

EshaanAgg commented on src/typeDefs/inputs.ts:
```diff
@@ -64,6 +64,15 @@ export const inputs = gql`
+    capacity: Int!
+    name: String!
+    description: String
+    organizationId: ID!
```
No need to provide this here.

EshaanAgg commented on src/typeDefs/types.ts:
```diff
@@ -298,6 +299,14 @@ export const types = gql`
+    description: String
+    imageUrl: URL
+  }
+
```
Add a field known as `organization` on the here, and implement a field level resolver for the same.

EshaanAgg commented on src/models/Organization.ts:
```diff
@@ -171,6 +174,12 @@ const organizationSchema = new Schema(
         ref: "CustomField",
       },
     ],
+    venues: [
+      {
+        type: Schema.Types.ObjectId,
+        ref: "Venue",
+      },
+    ],
```
No need to store this information on the organization model.

EshaanAgg commented on src/models/Venue.ts:
```diff
@@ -0,0 +1,44 @@
+    type: String,
+    required: false,
+  },
+});
```
Also store to which organization does this venue belong to here with a `organizationId` parameter.

EshaanAgg reviewed (changes requested):
Needs some changes in the schema and the typedefs. When the same is completed, the changes must be migrated to the resolvers and the queries. 

Good job, though.

wingman47 commented:
> Needs some changes in the schema and the typedefs. When the same is completed, the changes must be migrated to the resolvers and the queries.
> 
> Good job, though.
Thanks.
I have made the required changes. I will modify the tests once the changes are confirmed. I hope its fine!

EshaanAgg commented:
You also need to create a resolver for Organization on Venue type. The rest looks good.

wingman47 commented:
> You also need to create a resolver for Organization on Venue type. The rest looks good.

Should I create a separate query resolver for it because as of now there's no query that asks for venue id? Also, I thought we don't need to fetch organization information of a venue through venue id for the requested features as of now. I hope I didn't miss something. 😅

palisadoes commented:
That should be OK for now

@EshaanAgg @sumitra19jha Please take another look again

wingman47 commented:
@palisadoes I am getting formatting error in PR Workflow but I don't have any formatting issues in those files. I was modifying the tests for new changes.
![image](https://github.com/PalisadoesFoundation/talawa-api/assets/49762303/92382999-a52b-465c-82cd-7edd0d34ac85)

palisadoes commented:
Sometimes the workflow detects files you edited due to merge conflicts

wingman47 commented:
@palisadoes @EshaanAgg please review the pull request. All the required changes have been made and I have also added the tests.

EshaanAgg commented on schema.graphql:
```diff
@@ -211,6 +211,15 @@ input DonationWhereInput {
+  description: String
+  file: String
+  name: String
+  organizationId: ID!
```
No need for `organizationId` here.

wingman47 replied:
`organizationId` is being used here to check here if the user who wants to edit the venue is admin of the organization or not. Should I take any other approach?

EshaanAgg replied:
Fetch it from the database. The id is a unique identifier in itself. We can't let the user provide a faulty organizationId or change the organization itself in the venue.

EshaanAgg commented on schema.graphql:
```diff
@@ -211,6 +211,15 @@ input DonationWhereInput {
 }
 
+input EditVenueInput {
+  _id: ID!
```
Rename to `id`

EshaanAgg commented on schema.graphql:
```diff
@@ -1207,6 +1221,23 @@ type UsersConnectionResult {
 }
 
+type Venue {
+  _id: ID!
```
Rename to `id`

EshaanAgg reviewed (changes requested):
You also need to create a field level resolver for `Organization` on the `Venue` model

palisadoes commented:
Please fix the conflicting files

palisadoes commented:
You have a field named `organizationID` referencing an `organization`.  This negatively impacts readability. Please modify.

![image](https://github.com/PalisadoesFoundation/talawa-api/assets/16875803/7a29f9c4-2cc3-4f28-af26-5fbc13dd4830)

wingman47 commented:
> You have a field named `organizationID` referencing an `organization`. This negatively impacts readability. Please modify.
> 
> ![image](https://private-user-images.githubusercontent.com/16875803/305132289-7a29f9c4-2cc3-4f28-af26-5fbc13dd4830.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDgwMTU3NjEsIm5iZiI6MTcwODAxNTQ2MSwicGF0aCI6Ii8xNjg3NTgwMy8zMDUxMzIyODktN2EyOWY5YzQtMmNjMy00ZjI4LWFmMjYtNWZiYzEzZGQ0ODMwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMTUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjE1VDE2NDQyMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTg3ZGFlZDM0OTQ5ZDVmMjMzNzcwOGFhNjk2MDZhN2NhOGQ2NzU0ZWFkODBiNGQxODYxOGYyOTIzMGNlMGNkN2ImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.JOiMeJlA7V3pHFnHXPGbkPLO7sU87VMGlQLMIyFgbeU)
> 
> Then you have this, which adds to the inconsistency :
> 
> ![image](https://private-user-images.githubusercontent.com/16875803/305133317-1b3daf7b-74e1-4eb3-8730-3a677bdfffd8.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDgwMTU3NjEsIm5iZiI6MTcwODAxNTQ2MSwicGF0aCI6Ii8xNjg3NTgwMy8zMDUxMzMzMTctMWIzZGFmN2ItNzRlMS00ZWIzLTg3MzAtM2E2NzdiZGZmZmQ4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMTUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjE1VDE2NDQyMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTI3ZjhjNjE0Njc2NWMzNGY0NGMyMjRiYzFlM2FhOGE0N2UyMzg0N2ZiZmZhYzA2NjBlNzg5YWU4NGM2MTI2MDYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.uQUxZQZD6ygtKH9Q8UoQxD8Hyq-pVzvh7mMeJP6o5B4)

Ok, I will rename it to organization. Will it be okay?
Also, I am unable to write unit test for `updatedVenue is possibly null` in `editVenue` mutation because the `venueId` being using in `findOneAndUpdate` is already verified above so there's no possibillity of `updatedVenue` to be null and I am not allowed to use no null assertion for `updatedVenue` in return due to the workflow rules. Any suggestion to avoid this?

![image](https://github.com/PalisadoesFoundation/talawa-api/assets/49762303/f8f88ce6-645e-46b0-a3f5-9a7c7919d0a0)

sumitra19jha commented:
This branch is out-of-date with the base branch

wingman47 commented:
> This branch is out-of-date with the base branch

Done
I am unable to increase code cov due to above stated reason

palisadoes commented:
*NOTE* Read very carefully

1. We just merged this PR which upgraded the prettier package. 
    1. https://github.com/PalisadoesFoundation/talawa-api/pull/1840
1. It reformatted over 400 files. This will put your PR at risk of extensive merge conflicts.    
1. I suggest that before you merge your code to the latest upstream, or manually fix these merge conflicts that you:
    1. upgrade your `prettier` in your local branch to the same version
    1. run `prettier`
    
This will help to reduce the number of future merge conflicts for your PR.

wingman47 commented:
@palisadoes I have made all the requested changes.

palisadoes commented:
@EshaanAgg @sumitra19jha @SiddheshKukade 

PTAL

SiddheshKukade commented:
@wingman47 please take a pull from origin develop

wingman47 commented:
> @wingman47 please take a pull from origin develop

Done

EshaanAgg commented:
Please address all the previous conversations left in the previous reviews, and close the ones that you have addressed.

wingman47 commented:
> Please address all the previous conversations left in the previous reviews, and close the ones that you have addressed.

I have addressed all the reviews. 
Also, I am unable to cover a test case in edit Venue because of this reason:


> > You have a field named `organizationID` referencing an `organization`. This negatively impacts readability. Please modify.
> > ![image](https://private-user-images.githubusercontent.com/16875803/305132289-7a29f9c4-2cc3-4f28-af26-5fbc13dd4830.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDgwMTU3NjEsIm5iZiI6MTcwODAxNTQ2MSwicGF0aCI6Ii8xNjg3NTgwMy8zMDUxMzIyODktN2EyOWY5YzQtMmNjMy00ZjI4LWFmMjYtNWZiYzEzZGQ0ODMwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMTUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjE1VDE2NDQyMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTg3ZGFlZDM0OTQ5ZDVmMjMzNzcwOGFhNjk2MDZhN2NhOGQ2NzU0ZWFkODBiNGQxODYxOGYyOTIzMGNlMGNkN2ImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.JOiMeJlA7V3pHFnHXPGbkPLO7sU87VMGlQLMIyFgbeU)
> > Then you have this, which adds to the inconsistency :
> > ![image](https://private-user-images.githubusercontent.com/16875803/305133317-1b3daf7b-74e1-4eb3-8730-3a677bdfffd8.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDgwMTU3NjEsIm5iZiI6MTcwODAxNTQ2MSwicGF0aCI6Ii8xNjg3NTgwMy8zMDUxMzMzMTctMWIzZGFmN2ItNzRlMS00ZWIzLTg3MzAtM2E2NzdiZGZmZmQ4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMTUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjE1VDE2NDQyMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTI3ZjhjNjE0Njc2NWMzNGY0NGMyMjRiYzFlM2FhOGE0N2UyMzg0N2ZiZmZhYzA2NjBlNzg5YWU4NGM2MTI2MDYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.uQUxZQZD6ygtKH9Q8UoQxD8Hyq-pVzvh7mMeJP6o5B4)
> 
> Ok, I will rename it to organization. Will it be okay? Also, I am unable to write unit test for `updatedVenue is possibly null` in `editVenue` mutation because the `venueId` being using in `findOneAndUpdate` is already verified above so there's no possibillity of `updatedVenue` to be null and I am not allowed to use no null assertion for `updatedVenue` in return due to the workflow rules. Any suggestion to avoid this?
> 
> ![image](https://private-user-images.githubusercontent.com/49762303/305153778-f8f88ce6-645e-46b0-a3f5-9a7c7919d0a0.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDgzNzYzNTYsIm5iZiI6MTcwODM3NjA1NiwicGF0aCI6Ii80OTc2MjMwMy8zMDUxNTM3NzgtZjhmODhjZTYtNjQ1ZS00NmIwLWEzZjUtOWE3Yzc5MTlkMGEwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjE5VDIwNTQxNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWMyMjhkNjVlNmFjMmFkYTJjZTA0ZjE3ZDU2Y2RlOGNlMGI3NDMzYzQ1MGI4YzQzNjNjOWY3M2ZiNDE4ZjdlNWEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.5NrLVAAxwVSW6Ak0QvMlft7bKJrnHqNXGv6t615dojA)

Any suggestion to handle this?

wingman47 commented:
@EshaanAgg Please review the PR
